### PR TITLE
Circuit Read Iterator

### PR DESCRIPTION
**Pull request by peterschwarz**
Convert the vector to an iterator, which provides more flexibility for circuit store back-ends.

### CODE REVIEW

File: libsplinter/src/circuit/store.rs [hunk 1]
```diff
@@ -36,11 +36,46 @@ impl CircuitFilter {
 }
 
 pub trait CircuitStore: Send + Sync + Clone {
-    fn circuits(&self) -> Result<BTreeMap<String, Circuit>, CircuitStoreError>;
+    /// Return an iterator over the circuits in this store.
+    ///
+    /// A circuit filter may optionally provided, to reduces the results.
+    fn circuits(&self, filter: Option<CircuitFilter>) -> Result<Circuits, CircuitStoreError>;
```
Comment (vaporos) [inline libsplinter/src/circuit/store.rs]: circuit/circuits Circuit/Circuits are very closely named, maybe list_circuits()->CircuitList?

Comment (vaporos) [reply libsplinter/src/circuit/store.rs]: Should we just return Iterator<Item = Circuit> here so we avoid exposing the implementation completely?

Comment (peterschwarz) [reply libsplinter/src/circuit/store.rs]: Unfortunately, no.  I tried to go down that route with event iteration, and the compiler did not like it due to the async layer at the REST API level. We may be able to change that when we more cleanly separate the rest API business logic layer from the underlying http server library. 

It's not a list, though, so I don't think CircuitList isn't the right name.  This is also a pretty common pattern in the standard library:  See `String::chars() -> Chars` as a good example. I could probably be convinced to call it `CircuitIter`, though.

Comment (vaporos) [reply libsplinter/src/circuit/store.rs]: CircuitIter seems fine

Comment (peterschwarz) [reply libsplinter/src/circuit/store.rs]: Will add an new PR for this.

**Pull request merged**